File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/unit/AirQualityCOUnit.java
package wangdaye.com.geometricweather.common.basic.models.options.unit;

import android.content.Context;
import android.text.BidiFormatter;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;

public enum AirQualityCOUnit {

    MGPCUM("mgpcum", 0, 1f);

    private final String unitId;
    private final int unitArrayIndex;
    private final float unitFactor; // actual air quality CO = quality(mg/m³) * factor.

    AirQualityCOUnit(String id, int arrayIndex, float factor) {
        unitId = id;
        unitArrayIndex = arrayIndex;
        unitFactor = factor;
    }

    public String getUnitId() {
        return unitId;
    }

    public float getDensity(float mgpcum) {
        return mgpcum * unitFactor;
    }

    public String getDensityText(Context context, float mgpcum) {
        return getDensityText(context, mgpcum, DisplayUtils.isRtl(context));
    }

    private String getDensityText(Context context, float mgpcum, boolean rtl) {
        if (rtl) {
            return BidiFormatter.getInstance().unicodeWrap(
                    UnitUtils.formatFloat(mgpcum * unitFactor, 1)
            ) + "\u202f" + context.getResources().getStringArray(R.array.air_quality_co_units)[unitArrayIndex];
        } else {
            return UnitUtils.formatFloat(mgpcum * unitFactor, 1)
                    + "\u202f" +  context.getResources().getStringArray(R.array.air_quality_co_units)[unitArrayIndex];
        }
    }

    public String getDensityVoice(Context context, float mgpcum) {
        return getDensityVoice(context, mgpcum, DisplayUtils.isRtl(context));
    }

    private String getDensityVoice(Context context, float mgpcum, boolean rtl) {
        if (rtl) {
            return BidiFormatter.getInstance().unicodeWrap(
                    UnitUtils.formatFloat(mgpcum * unitFactor, 1)
            ) + "\u202f" +  context.getResources().getStringArray(R.array.air_quality_co_unit_voices)[unitArrayIndex];
        } else {
            return UnitUtils.formatFloat(mgpcum * unitFactor, 1)
                    + "\u202f" +  context.getResources().getStringArray(R.array.air_quality_co_unit_voices)[unitArrayIndex];
        }
    }
}

File: app/src/main/java/org/breezyweather/common/basic/models/options/unit/AirQualityCOUnit.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.basic.models.options.unit

import android.content.Context
import android.icu.util.MeasureUnit
import android.os.Build
import org.breezyweather.R
import org.breezyweather.common.basic.models.options.basic.UnitEnum
import org.breezyweather.common.basic.models.options.basic.UnitUtils

// actual air quality CO = quality(mg/m³) * factor.
enum class AirQualityCOUnit(
    override val id: String,
    override val measureUnit: MeasureUnit?,
    override val perMeasureUnit: MeasureUnit?,
    override val convertUnit: (Double) -> Double,
) : UnitEnum<Double> {

    MILLIGRAM_PER_CUBIC_METER(
        "mgpcum",
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) MeasureUnit.MILLIGRAM else null,
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) MeasureUnit.CUBIC_METER else null,
        { valueInDefaultUnit -> valueInDefaultUnit }
    ),
    ;

    override val valueArrayId = R.array.air_quality_co_unit_values
    override val nameArrayId = R.array.air_quality_co_units
    override val contentDescriptionArrayId = R.array.air_quality_co_unit_voices

    override fun getName(context: Context) = UnitUtils.getName(context, this)

    override fun getMeasureContentDescription(context: Context) = UnitUtils.getMeasureContentDescription(context, this)

    override fun getConvertedUnit(valueInDefaultUnit: Double) = convertUnit(valueInDefaultUnit)

    override fun formatValue(
        context: Context,
        valueInDefaultUnit: Double,
    ) = UnitUtils.formatValue(
        context = context,
        enum = this,
        value = valueInDefaultUnit,
        precision = 1
    )

    override fun formatMeasure(
        context: Context,
        value: Double,
        isValueInDefaultUnit: Boolean,
    ) = UnitUtils.formatMeasure(
        context = context,
        enum = this,
        value = value,
        precision = 1,
        isValueInDefaultUnit = isValueInDefaultUnit
    )

    override fun formatContentDescription(
        context: Context,
        value: Double,
        isValueInDefaultUnit: Boolean,
    ) = UnitUtils.formatMeasure(
        context = context,
        enum = this,
        value = value,
        precision = 1,
        isValueInDefaultUnit = isValueInDefaultUnit,
        unitWidth = UnitWidth.FULL
    )
}
